joomla/joomla-cms
# Adding the Multilanguage Associations Manager

# Executive summary
This PR is a partial rewrite of the com_associations component based ion the work done in the Google Summer of Code Project "Improved Multilingual" (Repo: https://github.com/joomla-projects/gsoc16_improved-multi-lingual)


## Background information
The original concept made a lot of assumptions about fields and extensions and instantiated extensions models to get information about the internal structure. This concept comes with two problems:

1. Guessing can be wrong :-)
2. Instantiate a model can end up in fatal errors and a not working Website when the extension has a bug.  

I removed the guess and model instantiating and replaced it with a help class for extensions. This helper class provides all the needed information.

# Backwards compatibility
It is a new function so no problem here

# Translation impact
Two new language files for com_associations and one new language for any core extension that supports associations.

# Testing instruction 

* clone or get a copy of the staging branch
* setup a multilingual website
* Apply patch
* discover install

You will now find a new extension under components. 
The select box allows to chose an item type.

![ml-pic1](https://cloud.githubusercontent.com/assets/467356/21805300/bdd3ae72-d733-11e6-8500-50e23f11afb9.png)

When you have chosen a type and language, you can see your articles. 

![ml-pic2](https://cloud.githubusercontent.com/assets/467356/21805313/ce25fdf2-d733-11e6-940b-4c3be29143dc.png)

Click on an article (reference) and you will see the reference article on the left side. On the right side you can chose a translation target and translate your article.

![ml-pic3](https://cloud.githubusercontent.com/assets/467356/21805326/dca6b2b8-d733-11e6-8575-71f375cb97b7.png)

If you click on the available translation you get directly the view with two articles side by side.

![ml-pic4](https://cloud.githubusercontent.com/assets/467356/21805337/eaa8791e-d733-11e6-9e87-0675eb454f2d.png)

# Developer documentation
If your extension support association and you will make your extension available in the Associations Manager you have to provide a helper class (we have examples for com_content, com_contact, com_newsfeeds, com_menus, please look at the example implementation for inspiration :-))

You need to create a file "associations.php" in your component admin helper directory. If your component is names foobar the help class must be named as "FoobarAssociationsHelper" and it must extend "JAssociationExtensionHelper". 

Define three protected vars:

$extension = 'com_foobar';
$itemTypes = array('your_itemname', 'category');
$associationsSupport = true;

Define three public functions:

public function getAssociations($typeName, $id)

This function provides an interface the "JLanguageAssociations::getAssociations" and returns an array of associated items 

public function getItem($typeName, $id)

This function should return a JTable Object for the type and id. Don't forget something like "JTable::addIncludePath(__DIR__ . '/../tables')" so that the class can be found 

public function getType($typeName = '')

Here you set up information per type, the function returns an array: 

array(
			'fields'  => $fields,
			'support' => $support,
			'tables'  => $tables,
			'joins'   => $joins,
			'title'   => $title
		);

For fields and support we have a template function that set defaults so the you only have to overwrite what is different in your extension. Make sure that $fields['title'] and $fields['state'] are set right.

We need to write some documentation in JDocs but I think that here should be enough as a start.


# Open Issues
* Ordering not great
* Deal with checkout items  
* Extensions with multiple use of categories (seems a theoretical problem, because we couldn't find an extension using it)

# Final words
I would like to says thanks to the GSOC project members you did most of the work.
